packages/backend/test/username.test.ts

```diff
@@ -0,0 +1,123 @@
+
+test.before(async (t: any) => {
+    const context = await startServer({ epochLength: EPOCH_LENGTH / 1000 })
+    Object.assign(t.context, context)
+})
+
+test('should set a username', async (t: any) => {
+    // sign up and sign in user
+    const { iden, commitment } = await signUp(t)
+    await signIn(t, commitment)
+
+    // first set a username
+    // pre-image by default is 0
+    await setUsername(t, iden, 0, 'initial-test-username123')
+
+    await new Promise((r) => setTimeout(r, EPOCH_LENGTH))
+
+    // execute the epoch transition
+    const prevEpoch = await t.context.unirep.currentEpoch()
+    await epochTransition(t)
+    for (;;) {
+        await new Promise((r) => setTimeout(r, 1000))
+        const findEpoch = await t.context.db.findOne('Epoch', {
+            where: { number: Number(prevEpoch) },
+        })
+        if (findEpoch) break
+    }
+
+    // user state transition
+    await userStateTransition(t, iden)
+
+    // change the username to something else
+    await setUsername(
+        t,
+        iden,
+        'initial-test-username123',
+        'second-test-username123'
+    )
+
+    t.pass()
+})
+
+test.serial(
+    'should fail to set the username that is already taken',
+    async (t: any) => {
+        // sign up and sign in user
+        const { iden, commitment } = await signUp(t)
+        await signIn(t, commitment)
+
+        // first set a username
+        // pre-image by default is 0
+        await setUsername(t, iden, 0, 'username123')
+
+        await new Promise((r) => setTimeout(r, EPOCH_LENGTH))
+
+        // execute the epoch transition
+        const prevEpoch = await t.context.unirep.currentEpoch()
+        await epochTransition(t)
+        for (;;) {
+            await new Promise((r) => setTimeout(r, 1000))
+            const findEpoch = await t.context.db.findOne('Epoch', {
+                where: { number: Number(prevEpoch) },
+            })
+            if (findEpoch) break
+        }
+
+        // user state transition
+        await userStateTransition(t, iden)
+
+        // try to change the username to the same one
+        await setSameUsername(t, iden, 'username123', 'username123')
+    }
+)
+
+test.serial('should fail to set with invalid proof', async (t: any) => {
+    // sign up and sign in user
+    const { iden, commitment } = await signUp(t)
+    await signIn(t, commitment)
+
+    // first set a username
+    // pre-image by default is 0
+    await setUsername(t, iden, 0, 'username456')
+
+    const { proof, publicSignals, blockNumber } = await genUsernameProof(
+        t,
+        iden,
+        0
+    )
+    await waitForBackendBlock(t, blockNumber)
+
+    // send a invalid proof
+    const r = await fetch(`${t.context.url}/api/usernames`, {
+        method: 'POST',
+        headers: {
+            'content-type': 'application/json',
+        },
+        body: JSON.stringify({
+            newUsername: 'username456',
+            publicSignals: [publicSignals, 0],
```

Comment:
I'd like to see a slightly cleaner format for invalid proofs. `publicSignals` is an array so writing `[publicSignals, 0]` creates an array with two entries, an array and the number 0. This could cause unrelated failures like accessing indexes out of range.

We want to test that the verifier is properly examining public signals and proof values.

Try something like `publicSignals: publicSignals.reverse()`